pkg/resolvers/k8s/watcher.go

```diff
@@ -54,119 +51,136 @@ func (w *watcher) Close() {
 	if w.ctx.Err() != nil {
 		// We already stopped.
-		return []*naming.Update(nil), errors.Wrap(w.ctx.Err(), "k8sresolver: watcher.Next already stopped or Next returned error already. "+
+		return []*naming.Update(nil), errors.Wrap(w.ctx.Err(), "watcher.Next already stopped or Next returned error already. "+
 			"Note that watcher errors are not recoverable.")
 	}
 	u, err := w.next()
 	if err != nil {
 		// Just in case.
 		w.Close()
+		return u, errors.Wrap(err, "k8sresolver")
+	}
+	return u, nil
+}
+
+type key struct {
+	// EndpointAddress fields (usually pod).
+	ns, name string
+}
+
+func keyFromAddr(address v1.EndpointAddress) (key, error) {
+	if address.TargetRef == nil {
+		return key{}, errors.New("address targetRef is empty. Cannot maintain internal state.")
 	}
-	return u, err
+	return key{ns: address.TargetRef.Namespace, name: address.TargetRef.Name}, nil
 }
 
 func (w *watcher) next() ([]*naming.Update, error) {
-	updates := make([]*naming.Update, 0)
-	updatedEndpoints := make(map[string]struct{})
-	var event event
+	var (
+		updates           []*naming.Update
+		change            change
+		changeCh, errCh   = w.streamer.ResultChans()
+		endpointsToUpdate = map[key]string{}
+	)
+
 	select {
 	case <-w.ctx.Done():
 		// We already stopped.
 		return []*naming.Update(nil), w.ctx.Err()
-	case r := <-w.watchChange:
-		if r.err != nil {
-			return []*naming.Update(nil), errors.Wrap(r.err, "k8sresolver: error on reading event stream")
-		}
-		event = *r.ep
+	case err := <-errCh:
+		return []*naming.Update(nil), errors.Wrap(err, "error on reading change stream")
+	case change = <-changeCh:
 	}
 
-	// Translate kube api endpoint watch event to resolver address and put into map for easier lookup.
-	for _, subset := range event.Object.Subsets {
-		updatedAddresses, err := subsetToAddresses(w.target, subset)
+	// Translate kube api endpoint watch change to resolver address and put into map for easier lookup.
+	for _, subset := range change.Subsets {
+		var err error
+		endpointsToUpdate, err = subsetToAddresses(w.target, subset)
 		if err != nil {
-			return []*naming.Update(nil), errors.Wrap(err, "k8sresolver: failed to convert k8s endpoint subset to update Addr")
+			return []*naming.Update(nil), errors.Wrap(err, "failed to convert k8s endpoint subset to update Addr")
 		}
 
-		for _, address := range updatedAddresses {
-			updatedEndpoints[address] = struct{}{}
+		if len(endpointsToUpdate) > 0 {
+			// Expected port found.
+			break
 		}
+
+		// Target port not found yet. Maybe other subsets includes target one?
 	}
 
-	// Create updates to add new endpoints.
-	for addr, md := range updatedEndpoints {
-		if _, ok := w.lastUpdates[addr]; ok {
-			continue
+	switch change.typ {
+	case watch.Added:
+		// Create updates to add new endpoints.
+		for k, addr := range endpointsToUpdate {
+			_, ok := w.endpoints[k]
+			if ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
```

Comment:
Why not use `errors.Wrapf` here?

Reply:
Again, am  I wrapping any error? IMO I am creating new one (: